packages/edgeworth/src/synthesis/Synthesizer.ts

```diff
@@ -444,7 +445,20 @@ export class Synthesizer {
       // reset synthesizer after generating each Substance diagram
       this.reset();
       return sub;
-    });
+    };
+
+    let temp = dedupSynthesizedSubstances(_.times(numProgs, iteratee));
+
+    // Generate until there are numProgs unique Substance programs
+    while (temp.length < numProgs) {
+      const sub = iteratee(temp.length);
+      temp.push(sub);
+      temp = dedupSynthesizedSubstances(temp);
+    }
+
+    const substances = temp;
+    return substances;
```

Comment:
```suggestion
    let substances = dedupSynthesizedSubstances(_.times(numProgs, iteratee));

    // Generate until there are numProgs unique Substance programs
    while (substances.length < numProgs) {
      const sub = oneSubstance(temp.length);
      substances.push(sub);
      substances = dedupSynthesizedSubstances(substances);
    }
    return substances;
```

Small refactoring for readability again.